protocol/op-supervisor-integration.md

```diff
@@ -0,0 +1,309 @@
+
+#### finalized (always cross)
+
+To query the latest finalized L2 block,
+simply binary search for the last L2 block fully derived from the finalized L1 block.
+This can simplify existing op-node functionality, and make finality more reliable.
+
+#### cross-safe
+
+To query whether an L2 block is `cross-safe`, binary search for the L2 block, to determine when (w.r.t. L1 block)
+it became cross-safe. If this L1 block exists and is canonical, then the L2 block is cross-safe.
+
+#### local-safe
+
+The op-nodes determine this for themselves, and exchange it with the op-supervisor. The op-supervisor includes the "local safe" increment in the entry, but it may not be promoted to "cross safe" until all dependencies are met.
+
+#### cross-unsafe
+
+Optimistically the supervisor may fetch the event-logs of unsafe blocks. Blocks with transitively safe dependencies may be marked as cross-unsafe.
+
+The cross-unsafe data is not persisted across restarts: this can be re-computed from the event-log DB quickly,
+and recomputing makes it much less complicated to stay consistent with the database.
+
+#### unsafe
+
+The sync-status exchange triggers the op-supervisor to fetch the event-logs of new unsafe blocks.
+
+If the new unsafe-block does not build on the previous unsafe block (parent-hash check), then the event-log DB needs to be rewound to the last blockhash checkpoint still consistent with the post-reorg L2 chain.
+This may not revert deeper than cross-safe, as the L1 chain is the source of truth for reversals that deep.
+The cross-safe will has to rewind first, and then when unsafe-blocks still “don’t fit” on the remaining unsafe data, the unsafe data reorg is triggered again, and can be rewound further.
+
+### Super-safe-DB format
+
+The DB is append-only, with truncation on rollback:
+this makes it easy to recover on corruption, and fast and efficient.
+
+The entries are fixed-size:
+
+```go
+<version> // 1 byte identifier of entry
+<derived-up-and-including-to L1 block number> // 8 bytes, primary key
+<derived-up-and-including-to L1 truncated blockhash> // 20 bytes
+
+<chain idx> // 4 bytes, same indexing as executing event-log entry
+
+<l2 local-safe number> // 8 bytes. local-safe will be promoted to cross-safe once all dependencies transitively verify
+<l2 local-safe truncated blockhash> // 20 bytes. May refer to an invalid non-canonical L2 block, as dependencies are unverified.
+
+<l2 cross-safe number> // 8 bytes. Each L2 has its own. Also implied to be finalized, if derived-from L1 is finalized.
+<l2 cross-safe truncated blockhash> // 20 bytes.
+
+<flags> // 1 byte, to indicate if cross-safe was a deposit-only block
+// total: 1+8+20+4+8+20+8+20+1 = 90 bytes
+```
+
+For each chain in the dependency set, for each L1 block,
+we append one of these entries to the super-safe-db.
+If an L2 chain has not traversed gotten to the next L1 block yet, the DB writing work halts until it does.
```

Comment:
So we have two invariants here:
1. L1 block numbers are fully sequential.  ie we write entries for every block even if there are no batches in that block and nothing changes on any chain.
2. The entries from each L2 for a specific L1 block are all adjacent to each other. If we're binary searching they probably need to be in the same order for each L1 block for simplicity though we could make it work if they're out of order.

Reply:
1. Exactly
2. Yes, L2 entries are grouped together. Within the group we could sort them by chainID. But binary-search over the L1 blocks would work the same; just need to get the data of the L2 that is being searched for from the group.

Reply:
I'm just thinking that a full ordering (ie sort the L2 entries) would mean you just do a binary search and it will find exactly the entry for the L2 you want.  Whereas if they aren't sorted you binary search to find the group, then have to scan backwards or forwards among those entries to find the one you want.  I guess its simpler if the binary search can always find the first entry from the group but given the possibility of adding or removing chains from the group I don't think it can.

Reply:
Yes, sorting within groups makes sense, then it's easier to find particular entries